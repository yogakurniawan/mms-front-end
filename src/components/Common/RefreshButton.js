import React from 'react';
import PropTypes from 'prop-types';
import FlatButton from 'material-ui/FlatButton';
import NavigationRefresh from 'material-ui/svg-icons/navigation/refresh';

const RefreshButton = ({ label, refresh }) => <FlatButton
    primary
    label={label && label}
    onClick={refresh}
    icon={<NavigationRefresh />}
/>;

RefreshButton.defaultProps = {
    refresh: () => {}
}

RefreshButton.propTypes = {
    label: PropTypes.string,
    refresh: PropTypes.func.isRequired
};

export default RefreshButton;
