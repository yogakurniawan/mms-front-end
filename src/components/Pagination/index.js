import React, { Component } from 'react';
import PropTypes from 'prop-types';
import FlatButton from 'material-ui/FlatButton';
import IconButton from 'material-ui/IconButton';
import ChevronLeft from 'material-ui/svg-icons/navigation/chevron-left';
import ChevronRight from 'material-ui/svg-icons/navigation/chevron-right';
import { Toolbar, ToolbarGroup } from 'material-ui/Toolbar';
import withWidth from 'material-ui/utils/withWidth';
import RefreshButton from 'components/Common/RefreshButton';

const styles = {
  button: {
    margin: '10px 0',
  },
  pageInfo: {
    padding: '1.2em',
  },
  mobileToolbar: {
    margin: 'auto',
  },
  toolbar: {
    backgroundColor: 'rgb(237, 236, 236)'
  }
};

class Pagination extends Component {
  range() {
    const input = [];
    const { page, perPage, total } = this.props;
    if (isNaN(page)) return input;
    const nbPages = Math.ceil(total / perPage) || 1;

    // display page links around the current page
    if (page > 2) {
      input.push('1');
    }
    if (page === 4) {
      input.push('2');
    }
    if (page > 4) {
      input.push('.');
    }
    if (page > 1) {
      input.push(page - 1);
    }
    input.push(page);
    if (page < nbPages) {
      input.push(page + 1);
    }
    if (page === (nbPages - 3)) {
      input.push(nbPages - 1);
    }
    if (page < (nbPages - 3)) {
      input.push('.');
    }
    if (page < (nbPages - 1)) {
      input.push(nbPages);
    }

    return input;
  }

  getNbPages() {
    return Math.ceil(this.props.total / this.props.perPage) || 1;
  }

  prevPage = (event) => {
    event.stopPropagation();
    if (this.props.page === 1) {
      throw new Error('navigation.page_out_from_begin');
    }
    this.props.setPage(this.props.page - 1);
  }

  nextPage = (event) => {
    event.stopPropagation();
    if (this.props.page > this.getNbPages()) {
      throw new Error('navigation.page_out_from_end');
    }
    this.props.setPage(this.props.page + 1);
  }

  gotoPage = (event) => {
    event.stopPropagation();
    const page = event.currentTarget.dataset.page;
    if (page < 1 || page > this.getNbPages()) {
      throw new Error('navigation.page_out_of_boundaries');
    }
    this.props.setPage(page);
  }

  renderPageNums() {
    return this.range().map((pageNum, index) =>
      (pageNum === '.') ?
        <span key={`hyphen_${index}`} style={{ padding: '1.2em' }}>&hellip;</span> :
        <FlatButton className="page-number" key={pageNum} label={pageNum} data-page={pageNum} onClick={this.gotoPage} primary={pageNum !== this.props.page} style={styles.button} />
    );
  }

  render() {
    const { page, perPage, total, width, refresh } = this.props;
    if (total === 0) return null;
    const offsetEnd = Math.min(page * perPage, total);
    const offsetBegin = Math.min((page - 1) * perPage + 1, offsetEnd);
    const nbPages = this.getNbPages();

    return width === 1 ? (
      <Toolbar style={styles.toolbar}>
        <ToolbarGroup style={styles.mobileToolbar}>
          {page > 1 &&
            <IconButton onClick={this.prevPage}>
              <ChevronLeft color='#6200C0' />
            </IconButton>
          }
          <span style={styles.pageInfo}>{`${offsetBegin}-${offsetEnd} of ${total}`}</span>
          {page !== nbPages &&
            <IconButton onClick={this.nextPage}>
              <ChevronRight color='#6200C0' />
            </IconButton>
          }
        </ToolbarGroup>
      </Toolbar>
    ) : (
        <Toolbar style={styles.toolbar}>
          <ToolbarGroup firstChild>
            <span className="displayed-records" style={styles.pageInfo}>{`${offsetBegin}-${offsetEnd} of ${total}`}</span>
            <RefreshButton refresh={refresh} label="Refresh" />
          </ToolbarGroup>
          {nbPages > 1 &&
            <ToolbarGroup>
              {page > 1 &&
                <FlatButton className="previous-page" primary key="prev" label="Prev" icon={<ChevronLeft />} onClick={this.prevPage} style={styles.button} />
              }
              {this.renderPageNums()}
              {page !== nbPages &&
                <FlatButton className="next-page" primary key="next" label="Next" icon={<ChevronRight />} labelPosition="before" onClick={this.nextPage} style={styles.button} />
              }
            </ToolbarGroup>
          }
        </Toolbar>
      );
  }
}

Pagination.propTypes = {
  page: PropTypes.number,
  perPage: PropTypes.number,
  total: PropTypes.number,
  width: PropTypes.number,
  setPage: PropTypes.func,
  refresh: PropTypes.func
};

export default withWidth()(Pagination);