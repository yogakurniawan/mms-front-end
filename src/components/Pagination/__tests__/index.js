import React from 'react';
import { mount } from 'enzyme';
import injectTapEventPlugin from 'react-tap-event-plugin';
import FlatButton from 'material-ui/FlatButton';
import IconButton from 'material-ui/IconButton';
import renderer from 'react-test-renderer';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import RefreshButton from 'components/Common/RefreshButton';
import Pagination from '../index';

injectTapEventPlugin();

describe('<Pagination />', () => {
  describe('mobile', () => {
    it('renders correctly', () => {
      const tree = renderer.create(
        <MuiThemeProvider>
          <Pagination
            refresh={jest.fn()}
            page={2}
            perPage={10}
            total={20}
            setPage={jest.fn()}
          />
        </MuiThemeProvider>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render a condensed <Toolbar>', () => {
      const wrapper = mount(<MuiThemeProvider>
        <Pagination page={2} perPage={5} total={15} width={1} />
      </MuiThemeProvider>);
      const iconButtons = wrapper.find(IconButton);
      expect(iconButtons.length).toEqual(2);
      const flatButtons = wrapper.find(FlatButton);
      expect(flatButtons.length).toEqual(0);
    });

    it('should render only the text when no pagination is necessary', () => {
      const page = 1;
      const perPage = 20;
      const total = 15;
      const offsetEnd = Math.min(page * perPage, total);
      const offsetBegin = Math.min((page - 1) * perPage + 1, offsetEnd);
      const wrapper = mount(
        <MuiThemeProvider>
          <Pagination page={page} perPage={perPage} total={total} width={1} />
        </MuiThemeProvider>);
      const iconButtons = wrapper.find(IconButton);
      expect(iconButtons.length).toEqual(0);
      const span = wrapper.find('span');
      expect(span.text()).toEqual(`${offsetBegin}-${offsetEnd} of ${total}`);
    });
  });

  describe('desktop', () => {
    it('should render a normal <Toolbar>', () => {
      const wrapper = mount(<MuiThemeProvider>
        <Pagination page={2} perPage={5} total={15} width={2} />
      </MuiThemeProvider>);
      const iconButtons = wrapper.find(IconButton);
      expect(iconButtons.length).toEqual(0);
      const flatButtons = wrapper.find(FlatButton);
      expect(flatButtons.length).toEqual(6);
    });

    it('should render a <RefreshButton>', () => {
      const wrapper = mount(<MuiThemeProvider>
        <Pagination page={2} perPage={5} total={15} width={2} />
      </MuiThemeProvider>);
      const refreshButton = wrapper.find(RefreshButton);
      expect(refreshButton.length).toEqual(1);
    });

    it('should render only the text when no pagination is necessary', () => {
      const page = 1;
      const perPage = 20;
      const total = 15;
      const offsetEnd = Math.min(page * perPage, total);
      const offsetBegin = Math.min((page - 1) * perPage + 1, offsetEnd);
      const wrapper = mount(<MuiThemeProvider>
        <Pagination page={page} perPage={perPage} total={total} width={2} />
      </MuiThemeProvider>);
      const flatButtons = wrapper.find(FlatButton);
      expect(flatButtons.length).toEqual(1);
      const span = wrapper.find('span.displayed-records');
      expect(span.text()).toEqual(`${offsetBegin}-${offsetEnd} of ${total}`);
    });
  });
});