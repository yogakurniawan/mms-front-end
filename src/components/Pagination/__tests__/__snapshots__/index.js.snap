exports[`<Pagination /> mobile renders correctly 1`] = `
<div
  className={undefined}
  style={
    Object {
      "WebkitTapHighlightColor": "rgba(0,0,0,0)",
      "backgroundColor": "rgb(237, 236, 236)",
      "boxSizing": "border-box",
      "display": "flex",
      "height": 56,
      "justifyContent": "space-between",
      "muiPrepared": true,
      "padding": "0px 24px",
    }
  }>
  <div
    className={undefined}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
        "marginLeft": -24,
        "marginRight": undefined,
        "muiPrepared": true,
        "position": "relative",
      }
    }>
    <span
      className="displayed-records"
      style={
        Object {
          "padding": "1.2em",
        }
      }>
      11-20 of 20
    </span>
    <button
      disabled={false}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "border": 10,
          "borderRadius": 2,
          "boxSizing": "border-box",
          "color": "#00bcd4",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "height": 36,
          "lineHeight": "36px",
          "margin": 0,
          "minWidth": 88,
          "muiPrepared": true,
          "outline": "none",
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "textAlign": "center",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
          "verticalAlign": null,
          "zIndex": 1,
        }
      }
      tabIndex={0}
      type="button">
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}>
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "inline-block",
              "fill": "#00bcd4",
              "height": 24,
              "marginLeft": 12,
              "marginRight": 0,
              "muiPrepared": true,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "verticalAlign": "middle",
              "width": 24,
            }
          }
          viewBox="0 0 24 24">
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z" />
        </svg>
        <span
          style={
            Object {
              "fontSize": 14,
              "fontWeight": 500,
              "letterSpacing": 0,
              "muiPrepared": true,
              "paddingLeft": 8,
              "paddingRight": 16,
              "position": "relative",
              "textTransform": "uppercase",
              "verticalAlign": "middle",
            }
          }>
          Refresh
        </span>
      </div>
    </button>
  </div>
  <div
    className={undefined}
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "space-between",
        "marginLeft": undefined,
        "marginRight": undefined,
        "muiPrepared": true,
        "position": "relative",
      }
    }>
    <button
      className="previous-page"
      disabled={false}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "border": 10,
          "borderRadius": 2,
          "boxSizing": "border-box",
          "color": "#00bcd4",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "height": 36,
          "lineHeight": "36px",
          "margin": "10px 0",
          "minWidth": 88,
          "muiPrepared": true,
          "outline": "none",
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "textAlign": "center",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
          "verticalAlign": null,
          "zIndex": 1,
        }
      }
      tabIndex={0}
      type="button">
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}>
        <svg
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          style={
            Object {
              "color": "rgba(0, 0, 0, 0.87)",
              "display": "inline-block",
              "fill": "#00bcd4",
              "height": 24,
              "marginLeft": 12,
              "marginRight": 0,
              "muiPrepared": true,
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "verticalAlign": "middle",
              "width": 24,
            }
          }
          viewBox="0 0 24 24">
          <path
            d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z" />
        </svg>
        <span
          style={
            Object {
              "fontSize": 14,
              "fontWeight": 500,
              "letterSpacing": 0,
              "muiPrepared": true,
              "paddingLeft": 8,
              "paddingRight": 16,
              "position": "relative",
              "textTransform": "uppercase",
              "verticalAlign": "middle",
            }
          }>
          Prev
        </span>
      </div>
    </button>
    <button
      className="page-number"
      data-page={1}
      disabled={false}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "border": 10,
          "borderRadius": 2,
          "boxSizing": "border-box",
          "color": "#00bcd4",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "height": 36,
          "lineHeight": "36px",
          "margin": "10px 0",
          "minWidth": 88,
          "muiPrepared": true,
          "outline": "none",
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "textAlign": "center",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
          "verticalAlign": null,
          "zIndex": 1,
        }
      }
      tabIndex={0}
      type="button">
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}>
        <span
          style={
            Object {
              "fontSize": 14,
              "fontWeight": 500,
              "letterSpacing": 0,
              "muiPrepared": true,
              "paddingLeft": 16,
              "paddingRight": 16,
              "position": "relative",
              "textTransform": "uppercase",
              "verticalAlign": "middle",
            }
          }>
          1
        </span>
      </div>
    </button>
    <button
      className="page-number"
      data-page={2}
      disabled={false}
      href={undefined}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={[Function]}
      onTouchTap={[Function]}
      style={
        Object {
          "WebkitTapHighlightColor": "rgba(0, 0, 0, 0)",
          "backgroundColor": "rgba(0, 0, 0, 0)",
          "border": 10,
          "borderRadius": 2,
          "boxSizing": "border-box",
          "color": "rgba(0, 0, 0, 0.87)",
          "cursor": "pointer",
          "display": "inline-block",
          "fontFamily": "Roboto, sans-serif",
          "fontSize": "inherit",
          "fontWeight": "inherit",
          "height": 36,
          "lineHeight": "36px",
          "margin": "10px 0",
          "minWidth": 88,
          "muiPrepared": true,
          "outline": "none",
          "overflow": "hidden",
          "padding": 0,
          "position": "relative",
          "textAlign": "center",
          "textDecoration": "none",
          "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          "userSelect": "none",
          "verticalAlign": null,
          "zIndex": 1,
        }
      }
      tabIndex={0}
      type="button">
      <div
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}>
        <span
          style={
            Object {
              "fontSize": 14,
              "fontWeight": 500,
              "letterSpacing": 0,
              "muiPrepared": true,
              "paddingLeft": 16,
              "paddingRight": 16,
              "position": "relative",
              "textTransform": "uppercase",
              "verticalAlign": "middle",
            }
          }>
          2
        </span>
      </div>
    </button>
  </div>
</div>
`;
